src/cls/rgw/cls_rgw.cc

```diff
@@ -3472,6 +3472,141 @@ static int rgw_cls_lc_get_head(cls_method_context_t hctx, bufferlist *in,  buffe
   return 0;
 }
 
+static int rgw_reshard_add(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  bufferlist::iterator in_iter = in->begin();
+
+  cls_rgw_reshard_add_op op;
+  try {
+    ::decode(op, in_iter);
+  } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: rgw_reshard_add: failed to decode entry\n");
+    return -EINVAL;
+  }
+
+  bufferlist bl;
+  ::encode(op.entry, bl);
+
+  string key = op.entry.bucket_name + "." + op.entry.bucket_id;
+  int ret = cls_cxx_map_set_val(hctx, key, &bl);
+  if (ret < 0) {
+    CLS_ERR("error adding reshard job for bucket %s with key %s",op.entry.bucket_name.c_str(), key.c_str());
+    return ret;
+  }
+
+  return ret;
+}
+
+static int rgw_reshard_list(cls_method_context_t hctx, bufferlist *in, bufferlist *out)
+{
+  cls_rgw_reshard_list_op op;
+  bufferlist::iterator in_iter = in->begin();
+  try {
+    ::decode(op, in_iter);
+  } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: rgw_cls_rehard_list(): failed to decode entry\n");
+    return -EINVAL;
+  }
+  cls_rgw_reshard_list_ret op_ret;
+  bufferlist::iterator iter;
+  map<string, bufferlist> vals;
+  string filter_prefix;
+  int ret = cls_cxx_map_get_vals(hctx, op.marker, filter_prefix, op.max, &vals);
+  if (ret < 0)
+    return ret;
+  map<string, bufferlist>::iterator it;
+  cls_rgw_reshard_entry entry;
+  for (it = vals.begin(); it != vals.end(); ++it) {
+    iter = it->second.begin();
+    try {
+    ::decode(entry, iter);
+    } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: rgw_cls_rehard_list(): failed to decode entry\n");
+    return -EIO;
+   }
+    op_ret.entries.push_back(entry);
+  }
+  op_ret.is_truncated = op.max && (op_ret.entries.size() >= op.max);
```

Comment:
need to pass `max + 1` to cls_cxx_map_get_vals() instead of `max` for this to work consistently.